rhine-bayes/app/Main.hs

```diff
@@ -0,0 +1,151 @@
+-- base
+import GHC.Float (float2Double, double2Float)
+
+-- transformers
+import Control.Monad.Trans.Class
+
+-- log-domain
+import Numeric.Log hiding (sum)
+
+-- monad-bayes
+import Control.Monad.Bayes.Class
+-- FIXME They should implement MMorph
+import Control.Monad.Bayes.Population hiding (hoist)
+import Control.Monad.Bayes.Sampler
+
+-- rhine
+import FRP.Rhine
+
+-- rhine-gloss
+import FRP.Rhine.Gloss.IO
+
+-- rhine-bayes
+import FRP.Rhine.Bayes
+import FRP.Rhine.Gloss.Common
+
+type StdDev = Double
+type Pos = (Double, Double)
+type Sensor = Pos
+
+-- | Harmonic oscillator with white noise
+prior1d :: (MonadSample m, Diff td ~ Double) =>
+  -- | Starting position
+  Double ->
+  -- | Starting velocity
+  Double ->
+  BehaviourF m td StdDev Double
+prior1d initialPosition initialVelocity = feedback 0 $ proc (stdDev, position') -> do
+  impulse <- arrM (normal 0) -< stdDev
+  -- FIXME make -3 input, sample once at the beginning, or on every key stroke
+  let acceleration = (-3) * position' + impulse
+  -- Integral over roughly the last 100 seconds, dying off exponentially, as to model a small friction term
+  velocity <- arr (+ initialPosition) <<< decayIntegral 100 -< acceleration
+  position <- integralFrom initialVelocity -< velocity
+  returnA -< (position, position)
+
+-- | 2D harmonic oscillator with noise
+prior :: (MonadSample m, Diff td ~ Double) => BehaviourF m td StdDev Pos
+prior = prior1d 10 0 &&& prior1d 0 10
+
+double2FloatTuple :: (Double, Double) -> (Float, Float)
+double2FloatTuple = double2Float *** double2Float
+
+decayIntegral :: (VectorSpace v (Diff td), Monad m) => Diff td -> BehaviourF m td v v
+decayIntegral timeConstant = average timeConstant >>> arr (timeConstant *^)
+
+noise :: MonadSample m => Behaviour m td Pos
+noise = let stdDev = 5 in whiteNoise stdDev &&& whiteNoise stdDev
+
+generativeModel :: (MonadSample m, Diff td ~ Double) => BehaviourF m td Pos Sensor
+generativeModel = proc latent -> do
+  noiseNow <- noise -< ()
+  returnA -< latent ^+^ noiseNow
+
+-- FIXME make parameters like n particles/spawn, softeq interactive!
+-- That means it'll not be a ClSF anymore
+
+model :: (MonadSample m, Diff td ~ Double) => BehaviourF m td StdDev (Sensor, Pos)
+model = proc stdDev -> do
+  latent <- prior -< stdDev
+  sensor <- generativeModel -< latent
+  returnA -< (sensor, latent)
+
+posterior :: (MonadInfer m, Diff td ~ Double) => BehaviourF m td (StdDev, Sensor) Pos
+posterior = bayesFilter model
```

Comment:
The `bayesFilter` concerns me a little. It looks like it relies on `SoftEq`, which in turn defaults to a particular observation model (namely normal with std=2). But that doesn't seem right (disregard if I'm missing something here).

Reply:
Yes, I think I need a bit of guidance here what typical nomenclature and conventions are. It's certain that we need some kind of importance function (is that the right name?), because exact equality immediately looses way to much probability mass. To make this configurable, I defined the `SoftEq` class, so one can configure the importance function via newtypes, while having hopefully sensible defaults for Double and so on. But I'm fuzzy on where it belongs. I used it here in the bayes filter, but that's a bit weird. I would have expected to use it in the particle filter directly (somewhere close to `runPopulation`), but couldn't find the right place.

Anyways, we should discuss this in `dunai-bayes`.

Reply:
https://github.com/turion/dunai-bayes/issues/3